tensorrt_llm/_torch/auto_deploy/transform/library/transformers_attn.py

```diff
@@ -0,0 +1,202 @@
+    value: torch.Tensor,
+    attention_mask: Optional[torch.Tensor],
+    dropout: float = 0.0,
+    scaling: Optional[float] = None,
+    **kwargs,
+):
+    """
+    A wrapper that accept HF attn inputs and outputs but calls flashinfer operator.
+    Used to patch attn_interface in transformers.
+    Args:
+        *args: same as HF attention_interface.
+        **kwargs: need to contain metadata, all_layers_kv_cache, and buffers from cm.
+    Returns:
+        attn_output: same as HF attention_interface.
+        attn_weights: set to None.
+    """
+
+    # First we extract cached attn inputs from kwargs
+    try:
+        metadata = kwargs["metadata"]
+        all_layers_kv_cache = kwargs["all_layers_kv_cache"]
+        buffers = kwargs["buffers"]
+    except KeyError as e:
+        raise KeyError(f"Missing required kwarg: {e.args[0]}")
+
+    # reshape to flashinfer input format
+    # TODO:(hg) might want to optimize this later.
+    query = torch.einsum("bhld->blhd", query)
+    key = torch.einsum("bhld->blhd", key)
+    value = torch.einsum("bhld->blhd", value)
+
+    # cm pass cache of all layers. we need to get the current layer's kv cache.
+    layer_idx = module._layer_idx
+    layer_kv_cache = (all_layers_kv_cache[layer_idx * 2], all_layers_kv_cache[layer_idx * 2 + 1])
+
+    attn_output = torch.ops.auto_deploy.flashinfer_attention_mha_with_cache(
+        # QKV
+        query,
+        key,
+        value,
+        # metadata
+        *metadata,
+        # caches
+        *layer_kv_cache,
+        # buffers
+        *buffers,
+        # constants, harded coded for flashinfer.
+        scaling,
+        k_scale=1,
+        v_scale=1,
+    )
+
+    attn_weights = None
+
+    return attn_output, attn_weights
+
+
+def forward_cached(module: torch.nn.Module, *cm_args, **cm_kwargs):
+    """
+    Run transformers module with cm inputs and cached attn.
+    Used to patch the original HFCausalLM.forward method.
+    """
+    expected_num_args = (
+        2 + module.num_cached_attn_args + module.num_kvcache_args + module.num_buffer_args
+    )
+    assert len(cm_args) == expected_num_args, (
+        f"Number of cm input args ({len(cm_args)}) mismatch. Expected {expected_num_args}."
+    )
+
+    input_ids, position_ids = cm_args[:2]
+
+    # cached_attn_args: seq_len, input_pos, cache_loc, pages_per_seq
+    cached_attn_args = cm_args[2 : 2 + module.num_cached_attn_args]
+    metadata = module.get_attn_metadata(input_ids, position_ids, *cached_attn_args)
+    assert len(metadata) == module.num_metadata, (
+        f"Number of metadata ({len(metadata)}) mismatch. Expected {module.num_metadata}."
+    )
+
+    # kv_cache_args: k_cache_0, v_cache_0, k_cache_1, v_cache_1, ...
+    kv_cache_args = cm_args[
+        2 + module.num_cached_attn_args : 2 + module.num_cached_attn_args + module.num_kvcache_args
+    ]
+
+    # buffers: (workspace_buffer,)
+    buffers = cm_args[2 + module.num_cached_attn_args + module.num_kvcache_args :]
```

Comment:
I don't like how we are extracting the per-layer set of arguments. 

My suggestion would be as follows:

1. For each layer we have an index map to collect the correct arguments from `cm_args`
2. We then have just have a list of list of arguments for each attention layer

Reply:
You can build the index map bowl in the `"transformers_replace_cached_attn"`